File: apps/agent/src/routes/graph.ts
import type { Request, Response, Router } from "express";
import { Router as createRouter } from "express";
import { z } from "zod";
import {
  graphStats,
  listEntities,
  listRelations,
  publicEntity,
  publicRelation,
  type EntityKind,
} from "../lib/graph.js";
import { runGraphExtraction } from "../agent/extract-graph.js";

export const graphRouter: Router = createRouter();

graphRouter.get("/graph", async (_req: Request, res: Response) => {
  try {
    const [stats, people, projects, topics, relations] = await Promise.all([
      graphStats(),
      listEntities({ kind: "person", limit: 60 }),
      listEntities({ kind: "project", limit: 30 }),
      listEntities({ kind: "topic", limit: 30 }),
      listRelations({ limit: 200 }),
    ]);
    return res.json({
      stats,
      entities: {
        person: people.map(publicEntity),
        project: projects.map(publicEntity),
        topic: topics.map(publicEntity),
      },
      relations: relations.map(publicRelation),
    });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return res.status(500).json({ error: "graph_load_failed", detail: msg });
  }
});

graphRouter.get("/graph/stats", async (_req: Request, res: Response) => {
  try {
    return res.json(await graphStats());
  } catch (err) {
    return res.status(500).json({ error: "stats_failed", detail: err instanceof Error ? err.message : String(err) });
  }
});

const entitiesQuery = z.object({
  kind: z.enum(["person", "project", "topic"]).optional(),
  limit: z.coerce.number().int().min(1).max(200).optional(),
});

graphRouter.get("/graph/entities", async (req: Request, res: Response) => {
  const parsed = entitiesQuery.safeParse(req.query);
  if (!parsed.success) {
    return res.status(400).json({ error: "invalid_query", detail: parsed.error.flatten() });
  }
  try {
    const opts: { kind?: EntityKind; limit?: number } = {};
    if (parsed.data.kind) opts.kind = parsed.data.kind;
    if (parsed.data.limit !== undefined) opts.limit = parsed.data.limit;
    const rows = await listEntities(opts);
    return res.json({ count: rows.length, entities: rows.map(publicEntity) });
  } catch (err) {
    return res.status(500).json({ error: "list_failed", detail: err instanceof Error ? err.message : String(err) });
  }
});

graphRouter.post("/graph/extract", async (req: Request, res: Response) => {
  const rebuild = req.body?.rebuild === true || req.query?.["rebuild"] === "1";

  res.status(200);
  res.setHeader("Content-Type", "text/event-stream");
  res.setHeader("Cache-Control", "no-cache, no-transform");
  res.setHeader("Connection", "keep-alive");
  res.setHeader("X-Accel-Buffering", "no");
  res.flushHeaders?.();

  let closed = false;
  res.on("close", () => { closed = true; });

  const sse = (event: string, data: Record<string, unknown>) => {
    if (closed) return;
    res.write(`event: ${event}\n`);
    res.write(`data: ${JSON.stringify(data)}\n\n`);
  };

  try {
    for await (const ev of runGraphExtraction({ rebuild })) {
      if (closed) break;
      sse(ev.kind, ev as unknown as Record<string, unknown>);
    }
  } catch (err) {
    sse("error", { message: err instanceof Error ? err.message : String(err), at: Date.now() });
  } finally {
    if (!closed) res.end();
  }
});

File: apps/agent-py/app/routes/graph.py
"""/graph — port of apps/agent/src/routes/graph.ts (read endpoints)."""
from __future__ import annotations

import json
from typing import Literal

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse, StreamingResponse

from app.agent.extract_graph import run_graph_extraction
from app.lib.graph import graph_stats, list_entities, list_relations, public_entity, public_relation

router = APIRouter()
_SSE_HEADERS = {"Cache-Control": "no-cache, no-transform", "Connection": "keep-alive", "X-Accel-Buffering": "no"}


@router.post("/graph/extract")
async def graph_extract(request: Request) -> StreamingResponse:
    body = {}
    try:
        body = await request.json()
    except Exception:  # noqa: BLE001
        pass
    rebuild = body.get("rebuild") is True or request.query_params.get("rebuild") == "1"

    async def gen():
        try:
            async for ev in run_graph_extraction(rebuild=rebuild):
                yield f"event: {ev['kind']}\ndata: {json.dumps(ev, default=str)}\n\n"
        except Exception as err:  # noqa: BLE001
            yield f"event: error\ndata: {json.dumps({'message': str(err)})}\n\n"

    return StreamingResponse(gen(), media_type="text/event-stream", headers=_SSE_HEADERS)


@router.get("/graph")
async def graph_route() -> JSONResponse:
    stats = await graph_stats()
    people = await list_entities("person", 60)
    projects = await list_entities("project", 30)
    topics = await list_entities("topic", 30)
    relations = await list_relations(200)
    return JSONResponse({
        "stats": stats,
        "entities": {
            "person": [public_entity(e) for e in people],
            "project": [public_entity(e) for e in projects],
            "topic": [public_entity(e) for e in topics],
        },
        "relations": [public_relation(r) for r in relations],
    })


@router.get("/graph/stats")
async def graph_stats_route() -> JSONResponse:
    return JSONResponse(await graph_stats())


@router.get("/graph/entities")
async def graph_entities_route(kind: Literal["person", "project", "topic"] | None = None,
                               limit: int | None = None) -> JSONResponse:
    rows = await list_entities(kind, limit or 200)
    return JSONResponse({"count": len(rows), "entities": [public_entity(e) for e in rows]})
